5_data_munge/src/munge_data_files.R

```diff
@@ -179,3 +179,63 @@ summarize_all_dat <- function(in_ind, out_file) {
 
   readr::write_csv(summary, out_file)
 }
+
+clean_sites <- function(in_ind, out_ind) {
+
+  proj_albers <- "+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs"
+
+  states <- st_as_sf(map("state", plot = FALSE, fill = TRUE)) %>%
+    bind_rows(mutate(st_as_sf(map('world', 'USA:Alaska', plot = FALSE, fill = TRUE)), ID = 'alaska'),
+              mutate(st_as_sf(map('world', 'USA:Hawaii', plot = FALSE, fill = TRUE)), ID = 'hawaii'),
+              mutate(st_as_sf(map('world', 'Puerto Rico', plot = FALSE, fill = TRUE)), ID = 'puerto rico')) %>%
+    st_transform(crs = proj_albers)
+
+  #states_with_buffer <- st_buffer(states, dist = 1)
+  # filter sites that are outside of the bounding box
+  sites <- readRDS(sc_retrieve(in_ind, 'getters.yml')) %>%
+    # filter to stream sites
+    filter(site_type %in% c('ST', 'ST-TS', 'ST-CA','ST-DCH', 'Spring', 'Stream', 'stream', '')) %>%
+    st_as_sf(coords = c('longitude', 'latitude'), crs = 4326) %>% #create sf object
+    st_transform(crs = proj_albers) %>%
+    st_join(states, join = st_within) # match sites that are within state polygons
+
+  # find path forward for those sites outside of polygons
+  sites_no_state <- filter(sites, is.na(ID))
+
+  # draw a bbox with small buffer for main geographic areas - lower 48, hawaii, alaska, PR
+  # puerto rico
+  puerto <- st_intersects(sites_no_state, st_buffer(st_as_sfc(st_bbox(states[52, ])), dist = 2000))
+  pot_puerto <- sites_no_state[which(!is.na(as.numeric(puerto))), ]
+
+  # alaska
+  alaska <- st_intersects(sites_no_state, st_buffer(st_as_sfc(st_bbox(states[50, ])), dist = 2000))
+  pot_alaska <- sites_no_state[which(!is.na(as.numeric(alaska))), ]
+  # hawaii
+  hawaii <- st_intersects(sites_no_state, st_buffer(st_as_sfc(st_bbox(states[51, ])), dist = 2000))
+  pot_hawaii <- sites_no_state[which(!is.na(as.numeric(hawaii))), ]
+  # lower 48
+  lower <- st_intersects(sites_no_state, st_buffer(st_as_sfc(st_bbox(states[1:49, ])), dist = 2000))
+  pot_lower <- sites_no_state[which(!is.na(as.numeric(lower))), ] %>% select(-ID)
+
+  # find nearest shape for each of these potential sites
+  # for out-states, we assume those to be the outstate
+  # for lower 48, we use nearest feature join
+  sites_fixed_ids <- mutate(pot_puerto, ID = 'puerto rico') %>%
+    bind_rows(mutate(pot_alaska, ID = 'alaska')) %>%
+    bind_rows(mutate(pot_hawaii, ID = 'hawaii')) %>%
+    bind_rows(st_join(x = pot_lower, y = states, join = st_nearest_feature))
+
+  # check sites that have no match
+  # some of these are in Samoan Islands, some appear to have the wrong longitude (high positive numbers)
+  # e.g., site 	21FLCOSP_WQX-45-03, if I turn the longitude to a negative value, lands in a stream in Florida
+  # just ignoring these for now
+  test <- filter(sites_no_state, !site_id %in% sites_fixed_ids$site_id)
+
+  sites_fixed <- filter(sites, !is.na(ID)) %>%
+    bind_rows(sites_fixed_ids) %>%
+    st_transform(crs = 4326)
```

Comment:
This function looks pretty smart to me! One question is how many sites there are that are pretty far from the nearest actual feature - out in the middle of the Pacific Ocean, for example, but still within AK's bounding box. Could you catch most of the unmatched but plausibly correct coordinates simply by putting a buffer around the original state/territory polygon rather than around its bounding box?

Reply:
Also, a buffer of 2000 appears to be pretty small relative to the expansion caused by the bbox:
```r
> states[1, ] %>% st_bbox() %>% st_as_sfc() %>% st_buffer(dist=2000) %>% plot(reset=FALSE)
> states[1, ] %>% st_buffer(dist=2000) %>% plot(add=TRUE)
> states[1, ] %>% plot(add=TRUE)
```
![image](https://user-images.githubusercontent.com/12039957/95620922-e0298680-0a3e-11eb-88d2-6b1d9c02e39f.png)

For comparison, here's a buffer of 20000 (keeping the same buffered bbox:
```r
> states[1, ] %>% st_bbox() %>% st_as_sfc() %>% st_buffer(dist=2000) %>% plot(reset=FALSE)
> states[1, ] %>% st_buffer(dist=20000) %>% plot(add=TRUE)
> states[1, ] %>% plot(add=TRUE)
```
![image](https://user-images.githubusercontent.com/12039957/95621043-12d37f00-0a3f-11eb-9356-a362d8d648a1.png)

Reply:
Yes, good point. At one point, I had a good reason to do the bbox, but I think I was coming across that issue when I was using distance to assign the region. So, I don't think it's a good reason any longer!